### DISCUSSION TITLE
Wall Connector API calls???

### DISCUSSION THREAD

**Discussion by nortonsk** [timdorr/tesla-api#760] [Q&A]
Anyone is capable to reverse engineer how is performing action using Tesla Mobile App into API calls?
![WhatsApp Image 2023-09-30 at 03 38 15](https://github.com/timdorr/tesla-api/assets/34988036/c6ef0e77-474f-4b61-a5bd-d05d0a18cf98)

* Start_charging a non tesla vehicle on the Tesla Wall Connector (button Start Charging)
* Show history of charging on the energy product...
* set  in Settings / Vehicle Charging - "Allow Charging at" and "Stop Charging by"

You have to have added a new energy product with : resource_type == "wall_connector"

Thank you for any leads or results.

**Comment by purcell-lab** [comment]
I see some new endpoints but nothing for control.

![Screenshot 2023-09-30 12 12 58](https://github.com/timdorr/tesla-api/assets/79175134/aff35c48-d417-4478-b778-f6e5cd1d817a)

**Reply by JuliensLab** [reply]
Can you explain how to get this response? Is this on the local network by polling the wall connector, or accessing Tesla's servers over the internet?

**Comment by nortonsk** [comment]
This applies the same for non-tesla vehicle as well for tesla vehicle when it is connected in outside of the charging schedule...
The only thing which is needs to be set is charging schedule: 
![WhatsApp Image 2023-10-04 at 13 09 28](https://github.com/timdorr/tesla-api/assets/34988036/66a594a1-8ee1-4f58-9e27-5893f1d7e8dc)


something like this:
![WhatsApp Image 2023-10-04 at 13 10 15](https://github.com/timdorr/tesla-api/assets/34988036/fcaa364d-3fc6-4983-9a29-e2c6f1b890e3)

### Anyone can help me to reverse engineer the API call from Tesla APP which is sending in order to perform:

1. GET HISTORY of charges (on the wall connector not the vehicle charges)
2. START CHARGING (issued for wall connector not for vehicle)
3. SET CHARGER SCHEDULE (for wall connector)
4. Any chance to SET CHARGING AMP - limit charging current based on Solar/Battery status?
BONUS: At the moment wall connector is able to report VIN of the Tesla car which is connected ... in case you can see if there is any identification for non tesla car it would be incredible.

That would be perfect to know.
In case someone can point me how to setup debugging of the android phone to show these info I would try to do that and share it here...

Thank you for help.

**Comment by JuliensLab** [comment]
Also looking for api access to control charging rate, start charge and stop charge.

**Comment by binou06** [comment]
Do someone found a solution ? 

When pushing button, it seems that this call is done (I obfuscate with XXXXXX what seems to be personnal)

`curl -H 'content-type: application/json' \
-H 'x-tesla-app-key: XXXXXXXXXXXXXXXXXX'  \
-H 'accept: */*'  \
-H 'authorization: Bearer <BEARER>'  \
-H 'x-app-state: Optional(active)' --compressed  \
-H 'accept-language: fr'  \
-H 'cache-control: no-cache'  \
-H 'charset: utf-8'  \
-H 'user-agent: Tesla/4.32.7 (com.teslamotors.TeslaApp; build:2649; iOS 17.4.1) Alamofire/5.2.1'  \
-H 'x-tesla-user-agent: TeslaApp/4.32.7-2649/XXXX/ios/17.4.1'  \
-H 'x-txid: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX'  \
-X POST https://owner-api.teslamotors.com/api/1/energy_sites/1689377630233112/command  \
-d '{"command_properties":{"message":{"wc":{"push_charge_command_request":{"charge_command":{"charging_command":1}}}},"identifier_type":4,"target_id":"XXXXXX-02-E--XXXXXXXX"},"command_type":"grpc_command"}'`

But when I try to replay it, the charge is not forced. Between two "real call" from the application, it seems than only the x-txid changes. Is this info is like a "signature" ? I think that this is the key of the problem. 

My problem is that in my region I have two "low cost period" from 1AM to 6AM and from 1PM to 5PM I can not configure in WC. I define only the night (to prevent the car from charging if I plug it during daytime). But the con is that I had a script in cron to force charge start at 1PM... it worked perfectly with my former (Wallbox Pulsar) but with my newly WC3 installed... it does not work anymore. Starting charge (on the car page of the application) does not override WC3 (I have to click on the "START CHARGING" button on the screenshot of this question).

If _x-txid_ means something to you, it is perhaps a solution ?
